### PR TITLE
cpu/atmega_common: Update to inlineable IRQ API

### PR DESCRIPTION
**Pull request by maribu**
### Contribution description

- Update IRQ API to the inline-able version
- Cleaned up the implementation
    - irq_disable() now is 2 CPU cycles instead of 4 CPU cycles
    - irq_restore() now is 3 CPU cycles (or 2 in the best case), instead of 5

### Testing procedure

1. Let Murdock check if everything still compiles
2. `make BOARD=arduino-mega2560 -C tests/irq_disable_restore flash test` (or any other ATmega based board)

### Issues/PRs references

Follow up of https://github.com/RIOT-OS/RIOT/pull/13999

### CODE REVIEW

**Comment by fjmolinas** [comment]
@maribu did you perform some benchmarks with `tests/bench_thread_yield_pingpong` for reference

**Comment by maribu** [comment]
I think I should split out the last commit into its own PR on rebase this one on top of that, once merged. But let's see if I can get Murdock happy first.

**Comment by maribu** [comment]
I split up some unrelated changes into their own commits. I will than bundle all clean ups that are not directly related to the goal of this PR into a separate PR. This should make review a bit easier.

**Comment by maribu** [comment]
Did by chance anyone start to look in the code? Otherwise, I would like to squash up-front.

**Comment by fjmolinas** [comment]
> Did by chance anyone start to look in the code? Otherwise, I would like to squash up-front.

Go ahead, I was planning on reviewing next week.

**Comment by benpicco** [comment]
This needs a rebase now

**Comment by benpicco** [comment]
Running `tests/irq_disable_restore` gives me

```
2020-05-18 17:46:37,495 # main(): This is RIOT! (Version: 2020.07-devel-615-gdb248-atmega_irq)
2020-05-18 17:46:37,529 # Test for irq_disable() / irq_restore()
2020-05-18 17:46:37,562 # ======================================
2020-05-18 17:46:37,563 # 
2020-05-18 17:46:37,613 # Verifying test works: [FAILURE]
2020-05-18 17:46:37,652 # Test result: [FAILURE]
2020-05-18 17:46:37,653 # Stack Pointer: 0x0635
2020-05-18 17:46:37,685 # *** RIOT kernel panic:
2020-05-18 17:46:37,711 # FAILED ASSERTION.
2020-05-18 17:46:37,711 # 
2020-05-18 17:46:37,712 # *** halted.
```

**Comment by maribu** [comment]
With the docker toolchain I can reproduce this. I'll investigate. Hopefully this is a real bug and not just the legacy toolchain.

**Comment by benpicco** [comment]
This is working now, with `tests/bench_thread_yield_pingpong` I see an improvement from 19524 ->  21992 (12.6%) and a reduction in `.text` size by 132 bytes.

Feel free to squash.

File: tests/irq_disable_restore/main.c [hunk 1]
```diff
@@ -25,7 +25,7 @@
 #include "irq.h"
 #include "xtimer.h"
 
-#define DELAY       (10 * US_PER_MS)
+#define DELAY       (50 * US_PER_MS)
```
Comment (benpicco) [inline tests/irq_disable_restore/main.c:28]: Why increase the delay?

Comment (maribu) [reply tests/irq_disable_restore/main.c:28]: Was my first try debugging. Let me undo

**Comment by maribu** [comment]
OK: With the additional memory barrier in place, the ROM grows by 150 bytes. The symbols `_xtimer_now64` and `_periph_timer_callback` have very different (and longer) assembly code. (`static inline _xtimer_now64()` is btw. not inlined; which really is a good thing on AVR. It is huge there.)

I still think that the added memory barrier is not really needed and just hides another bug. Let me do some more digging.

**Comment by fjmolinas** [comment]
> I still think that the added memory barrier is not really needed and just hides another bug. Let me do some more digging.

@maribu  your last 2 commits addressed what you wanted right?

**Comment by fjmolinas** [comment]
Tested with `tests/bench_thread_yield_pingpong` I still get the same performance improvement as @benpicco of ~12.6%.

`RIOT_VERSION= BUILD_IN_DOCKER=1 make BOARD=arduino-mega2560 -C examples/hello-world/ clean all -j3`

- master
```
   text    data     bss     dec     hex filename
   5134     266     901    6301    189d /data/riotbuild/riotbase/examples/hello-world/bin/arduino-mega2560/hello-world.elf
```

- pr
```
   text    data     bss     dec     hex filename
   4994     266     901    6161    1811 /data/riotbuild/riotbase/examples/hello-world/bin/arduino-mega2560/hello-world.elf
```

No change in data/bss, ROM decrease in ~140bytes.

**Comment by fjmolinas** [comment]
@maribu if you are happy with your changes please squash!

**Comment by maribu** [comment]
> @maribu your last 2 commits addressed what you wanted right?

Yes. I believe if the compiler sees an assembly snipped *A* and some C code *B* and *A* is not touching memory nor any of the registers *B* is accessing, the compiler is allowed to reorder *A* and *B*. And this is what I believe happened here. The code now should be bullet proof regarding reordering and I tested it successfully using the docker image toolchain and my local toolchain.

**Comment by fjmolinas** [comment]
GO!

**Comment by maribu** [comment]
Thanks for the reviews!

Now only ESP, RISC-V, MSP430, and MIPS left :-) Maybe we should have opened a tracking issue...

**Pull request merged**